src/passes/SimplifyLocals.cpp

```diff
@@ -411,6 +411,13 @@ struct SimplifyLocals
     if (set->isTee()) {
       return false;
     }
+    // We cannot move expression containing exnref.pops that are not enclosed in
+    // 'catch', because 'exnref.pop' should follow right after 'catch'.
+    if (EffectAnalyzer(
+          this->getPassOptions(), this->getModule()->features, set->value)
+          .danglingPop) {
```

Comment:
I think we call `invalidates()` which would do what we need automatically. that is, it's ok to sink if we don't cross over another expression. and so this check is not needed.

Reply:
No, it's not about invalidating side effects. What we crossed over in my test case `$pop-cannot-be-sinked` is `(i32.const 3)`, which should be OK to be reordered with anything without changing any side effects. But `pop` cannot be reordered even with `i32.const` because `pop` should follow right after `catch`. Deleting this code make that test fail.

Reply:
Oh, come to think of it, I guess I should change `invalidates()` method instead, so that it returns true when either side of expression is `pop`, regardless of what the other side is? Is this what you meant? I'll do that. Please let me know if this is not what you meant.

Reply:
Hmm, it's weird, we're already checking that in `invalidate`... I'll investigate.

Reply:
So the test code that's failing is here: (Here `local.set` should NOT be sinked)
```wast
(catch
  (local.set $0 (exnref.pop))
  (call $foo
    (i32.const 3)
    (local.get $0)
  )
)
```
So `invalidates` check is called [here](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/passes/SimplifyLocals.cpp#L393-L395), and it is only called `checkPost` returns true. If that has to be called for our `(local.set (exnref.pop))`, at this point `original` should be `(i32.const 3)`. And `sinkables` [here](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/passes/SimplifyLocals.cpp#L286) would contain `(local.set (exnref.pop))`. But [`checkPost`](https://github.com/WebAssembly/binaryen/blob/beafd4721df6233556dfa330342ad10de7c0b9f5/src/ir/effects.h#L234-L240) does not return true for `(i32.const 3)`, because it does not have any side effects or anything. I'm not very familiar with how SimplifyLocals works, so I might be misunderstanding something.

Reply:
Oh right, it doesn't just check invalidates, it incrementally calls `checkPre/Post`. So yes, how you did it is the only way.